packages/tailwindcss/src/index.ts

```diff
@@ -353,42 +353,23 @@ export async function compile(
 
   // Replace `@apply` rules with the actual utility classes.
   if (css.includes('@apply')) {
-    walk(ast, (node, { replaceWith }) => {
-      if (node.kind === 'rule' && node.selector[0] === '@' && node.selector.startsWith('@apply')) {
-        let candidates = node.selector
-          .slice(7 /* Ignore `@apply ` when parsing the selector */)
-          .trim()
-          .split(/\s+/g)
-
-        // Replace the `@apply` rule with the actual utility classes
-        {
-          // Parse the candidates to an AST that we can replace the `@apply` rule with.
-          let candidateAst = compileCandidates(candidates, designSystem, {
-            onInvalidCandidate: (candidate) => {
-              throw new Error(`Cannot apply unknown utility class: ${candidate}`)
-            },
-          }).astNodes
-
-          // Collect the nodes to insert in place of the `@apply` rule. When a
-          // rule was used, we want to insert its children instead of the rule
-          // because we don't want the wrapping selector.
-          let newNodes: AstNode[] = []
-          for (let candidateNode of candidateAst) {
-            if (candidateNode.kind === 'rule' && candidateNode.selector[0] !== '@') {
-              for (let child of candidateNode.nodes) {
-                newNodes.push(child)
-              }
-            } else {
-              newNodes.push(candidateNode)
-            }
-          }
-
-          replaceWith(newNodes)
-        }
-      }
-    })
+    substituteAtApply(ast, designSystem)
```

Comment:
We could technically re-inline this function. But we will have to split it up immediately once this is merged because it's necessary in this PR: #14114